File: R/Classification/SocialNetworkAds-RandomForest.R
# Import Data set 
df = read.csv("C:/Users/snapp/Data-Projects/Datasets/Social_Network_Ads.csv")
df = df[, 3:5]

# Encoding the Target Feature aas Factor 
df$Purchased = factor(df$Purchased, levels = c(0,1))

# Split Dataset into train test split 
library(caTools)
# Set random state for seeing same data 
set.seed(123)

split = sample.split(df$Purchased, SplitRatio = 0.75)
train_set = subset(df, split == TRUE)
test_set = subset(df, split == FALSE)

# Feature Scaling 
# Need Numeric Columns 
train_set[, 1:2] = scale(train_set[, 1:2]) 
test_set[, 1:2] = scale(test_set[, 1:2])

# Fit Random Forest To the training set and Predict the test set results  
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = train_set[-3],
                          y =train_set$Purchased,
                          ntree = 500)

# Predict Test Results 
y_pred = predict(classifier, newdata = test_set[-3])

# Make Confusion Matrix 
CM = table(test_set[,3], y_pred)
print(CM)

# Visualising the Training Set Results 
library(ElemStatLearn)
set = train_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = predict(classifier, newdata = grid_set)
plot(set[, -3],
     main = 'Random Forest (Training set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, '#FF9999', '#99FF99'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, '#990000', '#009900'))

# Visualising the Test set results
library(ElemStatLearn)
set = test_set
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('Age', 'EstimatedSalary')
y_grid = predict(classifier, newdata = grid_set)
plot(set[, -3],
     main = 'Random Forest (Test set)',
     xlab = 'Age', ylab = 'Estimated Salary',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 1, '#FF9999', '#99FF99'))
points(set, pch = 21, bg = ifelse(set[, 3] == 1, '#990000', '#009900'))

File: Python/Classification/SocialNetworkAds-RandomForest.py
# Import Libraries 
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix

# Import the Dataset 
df = pd.read_csv(r"C:\Users\snapp\Data-Projects\Datasets\Social_Network_Ads.csv")
X = df.iloc[:,[2,3]].values 
y = df.iloc[:,-1].values

# Split Data into Train and Test set 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

# Feature Scaling 
# In Random Forest because it is not based on Euclidean distance it is not necessary to do  Feature scaling
# But we keep it for the Visualization 
SC_X = StandardScaler()
X_train = SC_X.fit_transform(X_train)
X_test = SC_X.transform(X_test)

# Fit Random Forest to the Training set 
classifier = RandomForestClassifier(n_estimators = 10, criterion= 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Predict the Test Set Results 
y_pred = classifier.predict(X_test)

# Confusion Matrix 
cm = confusion_matrix(y_test, y_pred)

# Visualise the Train set Results 
from matplotlib.colors import ListedColormap
X_set, y_set = X_train, y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:,0].min()-1, stop = X_set[:,0].max()+1, step = 0.01),
                     np.arange(start = X_set[:,1].min()-1, stop = X_set[:,1].max()+1, step = 0.01))

plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alph = 0.25, cmap = ListedColormap(('#FF9999', '#99FF99')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i , j  in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('#990000', '#009900'))(i), label = j, s = 5)
plt.title('Random Forest ')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()

# Visualise the Test set Results 
from matplotlib.colors import ListedColormap
X_set, y_set = X_test, y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:,0].min()-1, stop = X_set[:,0].max()+1, step = 0.01),
                     np.arange(start = X_set[:,1].min()-1, stop = X_set[:,1].max()+1, step = 0.01))

plt.contourf(X1, X2, classifier.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alph = 0.25, cmap = ListedColormap(('#FF9999', '#99FF99')))
plt.xlim(X1.min(),X1.max())
plt.ylim(X2.min(),X2.max())
for i , j  in enumerate(np.unique(y_set)):
    plt.scatter(X_set[y_set == j, 0], X_set[y_set == j, 1],
                c = ListedColormap(('#990000', '#009900'))(i), label = j, s = 5)
plt.title('Random Forest ')
plt.xlabel('Age')
plt.ylabel('Estimated Salary')
plt.legend()
plt.show()















